base_local_planner/src/speed_limiters/obstacle_speed_limiter.cpp

```diff
@@ -123,52 +134,41 @@ double SpeedLimiter::calculateAllowedLinearSpeed(costmap_2d::ObstructionMsg obs)
     abs_x_dist = obs.x - footprint_max_x_;
   }
 
-  double x_dist_with_buffer = std::max(0.0, abs_x_dist - params_.x_buffer_);
-  double y_dist_with_buffer = std::max(0.0, abs_y_dist - params_.y_buffer_);
+  double x_dist_with_buffer = std::max(0.0, abs_x_dist - params_.x_buffer);
+  double y_dist_with_buffer = std::max(0.0, abs_y_dist - params_.y_buffer);
 
 
   double distance_to_obstruction = std::sqrt(x_dist_with_buffer * x_dist_with_buffer + y_dist_with_buffer * y_dist_with_buffer);
   ROS_DEBUG("Obs: %f, %f.  abs x: %f, abs y: %f, Dist: %f", obs.x, obs.y, abs_x_dist, abs_y_dist, distance_to_obstruction);
 
-  double distance_scalar = std::max(0.0,
-    std::min(1.0,
-      (distance_to_obstruction - min_distance_to_stop_) / (max_distance_to_stop_ - min_distance_to_stop_)));
-
-  return params_.min_linear_velocity_ + (params_.max_linear_velocity_ - params_.min_linear_velocity_) * distance_scalar;
+  return threeLevelInterpolation(distance_to_obstruction,     
+    params_.min_range, params_.nominal_range_min,
+    params_.nominal_range_max, params_.max_range,
+    std::min(params_.min_linear_velocity, max_linear_velocity_),
+    std::min(params_.nominal_linear_velocity, max_linear_velocity_),
+    max_linear_velocity_);
 }
 
-double SpeedLimiter::calculateAllowedAngularSpeed(costmap_2d::ObstructionMsg obs)
-{
-  double distance_to_obstruction = std::sqrt(obs.x * obs.x + obs.y * obs.y) - circumscribed_radius_;
 
-  double distance_scalar = std::max(0.0,
-    std::min(1.0,
-      (distance_to_obstruction - params_.min_angular_velocity_effect_distance_)
-      / (params_.max_angular_velocity_effect_distance_ - params_.min_angular_velocity_effect_distance_)));
-
-  return params_.min_angular_velocity_ + (params_.max_angular_velocity_ - params_.min_angular_velocity_) * distance_scalar;
-}
-
-void SpeedLimiter::setCurrentPose(tf::Stamped<tf::Pose> pose)
+double ObstacleSpeedLimiter::calculateAllowedAngularSpeed(costmap_2d::ObstructionMsg obs)
 {
-  if (!pose.frame_id_.empty()
-    && world_frame_id_ != pose.frame_id_)
-  {
-    ROS_WARN("Received pose in frame %s, but world frame is %s",
-      pose.frame_id_.c_str(), world_frame_id_.c_str());
-  }
+  double distance_to_obstruction = std::sqrt(obs.x * obs.x + obs.y * obs.y) - circumscribed_radius_;
 
-  current_pose_inv_tf_ = pose.inverse();
+  return twoLevelInterpolation(distance_to_obstruction, 
+    params_.min_angular_velocity_effect_distance, params_.max_angular_velocity_effect_distance,
+    std::min(params_.min_angular_velocity, max_angular_velocity_), 
+    max_angular_velocity_);
 }
 
 
-costmap_2d::ObstructionMsg SpeedLimiter::obstructionToBodyFrame(const costmap_2d::ObstructionMsg& in)
+costmap_2d::ObstructionMsg ObstacleSpeedLimiter::obstructionToBodyFrame(const costmap_2d::ObstructionMsg& in,
+  tf::Pose current_pose_inv_tf)
```

Comment:
shouldn't current_pose_inv_tf also be a const ref?

Reply:
Done